repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -55,519 +54,363 @@ class KeystoneV2Filter @Inject()(configurationService: ConfigurationService,
-
-        // Construct and add impersonator roles, if available
-        val impersonatorRoles = token.impersonatorRoles
-        if (impersonatorRoles.nonEmpty) {
-          request.addHeader(OpenStackServiceHeader.IMPERSONATOR_ROLES, impersonatorRoles.mkString(","))
-        }
-
-        // Construct and add the tenant id header
-        val tenantsToPass = buildTenantHeader(token.defaultTenantId, token.tenantIds, matchedUriTenant)
-        if (tenantsToPass.nonEmpty) {
-          request.addHeader(OpenStackServiceHeader.TENANT_ID, tenantsToPass.mkString(","))
-        }
-
-        // If configured, add roles header
-        if (config.getIdentityService.isSetRolesInHeader && token.roles.nonEmpty) {
-          request.addHeader(OpenStackServiceHeader.ROLES, token.roles.map(_.name).mkString(","))
-        }
-
-        // If present, add the tenant from the URI as part of the Proxy header, otherwise use the default tenant id
-        matchedUriTenant match {
-          case Some(uriTenant) =>
-            request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION, s"$XAuthProxy $uriTenant")
-          case None =>
-            token.defaultTenantId match {
-              case Some(tenant) =>
-                request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION, s"$XAuthProxy $tenant")
-              case None =>
-                request.addHeader(OpenStackServiceHeader.EXTENDED_AUTHORIZATION, XAuthProxy)
-            }
-        }
-
-        // Construct and add authenticatedBy, if available
-        token.authenticatedBy foreach {
-          _ foreach {
-            request.addHeader(OpenStackServiceHeader.AUTHENTICATED_BY, _)
-          }
-        }
+    def addIdentityStatusHeader(confirmed: Boolean): Unit = {
+      if (Option(configuration.getDelegating).isDefined) {
+        if (confirmed) request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.CONFIRMED)
+        else request.addHeader(OpenStackServiceHeader.IDENTITY_STATUS, IdentityStatus.INDETERMINATE)
       }
+    }
 
-      def addCatalogHeader(maybeEndpoints: => Try[EndpointsData]): Try[Unit.type] = {
-        if (config.getIdentityService.isSetCatalogInHeader) {
-          maybeEndpoints map { endpoints =>
-            request.addHeader(PowerApiHeader.X_CATALOG, Base64.encodeBase64String(endpoints.json.getBytes))
-            Unit
-          }
-        } else {
-          Success(Unit)
+    /**
+      * BEGIN PROCESSING
+      */
+    getAuthToken flatMap { authToken =>
+      validateToken(authToken) flatMap { validToken =>
+        lazy val endpoints = getEndpoints(authToken) // Prevents making call if its not needed
+        val authResult = KeystoneV2Authorization.doAuthorization(configuration, request, validToken, endpoints)
+
+        addTokenHeaders(authResult.scopedToken, authResult.matchedTenant)
+        authResult match {
+          case AuthorizationPassed(scopedToken, _) =>
+            addCatalogHeader(endpoints)
```

Comment:
Hm, we are losing the return value from this call. Since `endpoints` is lazy, we might end up making the endpoints call here, so we need to capture the failure if that call fails. I suppose we could go back to nesting these calls to fix that.

```
addCatalogHeader(endpoints) flatMap { _ =>
  addGroupsHeader(getGroups(authToken, scopedToken))
}
```

Reply:
Lol, this was your suggested change

Reply:
I think they were side-by-side before my suggestion, though admittedly, I did not catch this when I made that suggestion.